### PR TITLE
Feature/647 value dist reducer

### PR DESCRIPTION
**Pull request by tomaszguzialek**
Ref #647 

This PR is competitive to #674, so have a look at both and merge only one!

I am posting this PR to gather feedback if I am on the right track and what I forgot to take care about. Especially Single vs GroupedValueDistributionResult.

The PR also adds a special test of reducing it in Spark + the new parameter to specify minimum number of partitions for the input data (in order to enforce more than one result that can be reduced later).

Also, ValueDistribution is not non-distributable anymore, so now using UniqueKeyCheck to test non-distributable jobs in Spark.

### CODE REVIEW

File: components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    @Inject
+    @Provided
+    RowAnnotationFactory _rowAnnotationFactory;
+
+    @Override
+    public ValueDistributionAnalyzerResult reduce(Collection<? extends ValueDistributionAnalyzerResult> analyzerResults) {
+        final ValueCountList topValues = ValueCountListImpl.emptyList();
+        final Collection<String> uniqueValues = Collections.emptyList();
+        final Map<String, RowAnnotation> annotations = Collections.emptyMap();
+        final InputColumn<?>[] highlightedColumns = null;
```
Comment (tomaszguzialek) [inline components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java]: I was not sure what to use as the annotations, RowAnnotationFactory, highlighted columns as the first value and later as a reduced value. Feedback welcome here.

File: components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+        SingleValueDistributionResult reducedResult = new SingleValueDistributionResult(null, topValues, uniqueValues,
+                0, 0, 0, annotations, new RowAnnotationImpl(), _rowAnnotationFactory, highlightedColumns);
+
+        for (ValueDistributionAnalyzerResult partialResult : analyzerResults) {
+            if (partialResult instanceof SingleValueDistributionResult) {
+                final SingleValueDistributionResult singleResult = (SingleValueDistributionResult) partialResult;
+                reducedResult = reduceSingleResult(annotations, highlightedColumns, reducedResult, singleResult);
```
Comment (kaspersorensen) [inline components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java]: I see a pattern here where we reduce only two results at a time. That might cause quite some performance overhead because of moving data out and around between many results. Would be better if the reducer would build one big result in a iterative way based on each result in a for-loop.

File: components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    private ValueCountList reduceTopValues(ValueCountList topValues1, ValueCountList topValues2) {
+        if (topValues1.getValueCounts().isEmpty()) {
+            return topValues2;
+        }
+
+        ValueCountListImpl topValuesImpl1 = (ValueCountListImpl) topValues1;
+        ValueCountListImpl topValuesImpl2 = (ValueCountListImpl) topValues2;
+
+        for (ValueFrequency valueFrequency2 : topValuesImpl2.getValueCounts()) {
+            for (ValueFrequency valueFrequency1 : topValuesImpl1.getValueCounts()) {
```
Comment (kaspersorensen) [inline components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java]: Many caveats here:

 * If there are unique values it seems that these get added to getValueCounts() as a single entry with key LabelUtils.UNIQUE_LABEL.
 * Since you are running two for-loops inside each other the performance characteristic here will be something like Big O = O^2. I think it should be doable to do it with a Big O = 2*O if you just append each value to a single new result.

**Comment by tomaszguzialek** [comment]
Ready for review again, but have a look at a competitive PR: #674

**Comment by tomaszguzialek** [comment]
Closing as #674 is preferred.
